community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java

```diff
@@ -969,7 +969,10 @@ public Property nodeSetProperty( KernelStatement state, long nodeId, DefinedProp
             // This because there are established ways of rebuilding auto-indexes which involves this function.
             autoIndexProperty( nodeId, property, ops, existingProperty, autoIndexing.nodes() );
 
-            if ( !property.equals( existingProperty ) )
+            //It is not enough to check equality here since by our equality semantics `int == tofloat(int)` is `true`
+            //so by only checking for equality users cannot change type of property without also "changing" the value.
+            //Hence the extra type check here.
+            if ( property.getClass() != existingProperty.getClass() || !property.equals( existingProperty ) )
```

Comment:
instead of copypasting comment and update check in 2 places please extract helper method instead